src/main/java/com/code_intelligence/jazzer/runtime/CoverageMap.java

```diff
@@ -96,9 +99,68 @@ public static void enlargeIfNeeded(int nextId) {
     }
   }
 
+  private static ConcurrentHashMap<Integer, Integer> edgeMap = new ConcurrentHashMap<Integer, Integer>();
+  private static int nextEdgeId = 0;
+  static ReentrantLock rl = new ReentrantLock();
+  private static Integer initializeEdgeForAndroid(int id) throws InterruptedException {
+    // while loop to run until id gets lock
```

Comment:
Could you explain what this map is used for? It does seem to map Jazzer's edge IDs obtained from offline instrumentation to consecutive IDs based on the order in which the edges are encountered at runtime. Is that really necessary? It seems that offline instrumentation could just emit globally unique IDs that Jazzer could use, possible enlargening the coverage map if an ID it encounters is larger than the current size.

Even if this remapping is required, I don't see why a lock around the `ConcurrentHashMap` is needed. It looks like `ConcurrentHashMap#computeIfAbsent` combined with `AtomicInteger#getAndIncrement` could be sufficient (and potentially much more performant and simple).

Reply:
You're right this re-map might not be needed. It is to keep the shared array with LibFuzzer smaller, since there can be some very large .jar files when fuzzing the Android Java framework. I didn't want to have a fuzz target that hits maybe 400 edges but has an array of size 2500,000 because it hit edge ID 250000 first.

Maybe the goal here should be to create smaller fuzz targets though rather than remapping the IDs.

Reply:
A large coverage map by itself shouldn't hurt performance, but instrumenting low-level system library code that runs in critical paths might. Preventing that would be the job of an appropriate Instrumentation filter though, so I think we can drop this remapping logic for now.